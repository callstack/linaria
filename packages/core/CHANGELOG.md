# Change Log

## 4.5.2

### Patch Changes

- Updated dependencies [85e74df6]
- Updated dependencies [1bf5c5b8]
  - @linaria/utils@4.5.1
  - @linaria/tags@4.5.2

## 4.5.1

### Patch Changes

- Updated dependencies [ceca1611]
- Updated dependencies [13258306]
  - @linaria/tags@4.5.1

## 4.5.0

### Minor Changes

- 16c057df: Breaking Change: Performance Optimization for `styled`

  When a component is wrapped in `styled`, Linaria needs to determine if that component is already a styled component. To accomplish this, the wrapped component is included in the list of variables for evaluation, along with the interpolated values used in styles. The issue arises when a wrapped component, even if it is not styled, brings along a substantial dependency tree. This situation is particularly evident when using `styled` to style components from third-party UI libraries.

  To address this problem, Linaria will now examine the import location of the component and check if there is an annotation in the `package.json` file of the package containing the components. This annotation indicates whether the package includes other Linaria components. If there is no such annotation, Linaria will refrain from evaluating the component.

  Please note that this Breaking Change solely affects developers of component libraries. In order for users to style components from your library, you must include the `linaria.components` property in the library's `package.json` file. This property should have a mask that covers all imported files with components. Here's an example of how to specify it:

  ```json
  "linaria": {
    "components": "**/*"
  }
  ```

### Patch Changes

- af5bb92d: The end of support for Node.js 14. Migration to pnpm 8.
- Updated dependencies [890b4aca]
- Updated dependencies [05ad266c]
- Updated dependencies [16c057df]
- Updated dependencies [af5bb92d]
  - @linaria/utils@4.5.0
  - @linaria/tags@4.5.0
  - @linaria/logger@4.5.0

## 4.2.10

### Patch Changes

- Updated dependencies [54ab61b2]
  - @linaria/tags@4.3.5
  - @linaria/utils@4.3.4

## 4.2.9

### Patch Changes

- 34029088: Usages of `styled` and `css` in Jest no longer trigger the "Using the â€¦ tag in runtime is not supported" exception.
- Updated dependencies [2e966f23]
- Updated dependencies [1c3f309d]
- Updated dependencies [dbe250b5]
  - @linaria/tags@4.3.4
  - @linaria/utils@4.3.3

## 4.2.8

### Patch Changes

- Updated dependencies [a3ad617f]
  - @linaria/tags@4.3.3

## 4.2.7

### Patch Changes

- Updated dependencies [f9df4ed8]
  - @linaria/utils@4.3.2
  - @linaria/tags@4.3.2

## 4.2.6

### Patch Changes

- Updated dependencies [28f3f93d]
- Updated dependencies [71a5b351]
- Updated dependencies [61d49a39]
  - @linaria/tags@4.3.1
  - @linaria/utils@4.3.1

## 4.2.5

### Patch Changes

- Updated dependencies [3ce985e0]
- Updated dependencies [d11174d0]
  - @linaria/tags@4.3.0
  - @linaria/utils@4.3.0

## 4.2.4

### Patch Changes

- Updated dependencies [315f0366]
  - @linaria/utils@4.2.6
  - @linaria/tags@4.2.2

## 4.2.3

### Patch Changes

- 5edde648: Upgrade Babel to support TypeScript 4.9. Fixes #1133.
- Updated dependencies [5edde648]
- Updated dependencies [b9e49b74]
  - @linaria/tags@4.2.1
  - @linaria/utils@4.2.5

## 4.2.2

### Patch Changes

- Updated dependencies [63f56d47]
- Updated dependencies [963508a2]
  - @linaria/tags@4.2.0
  - @linaria/utils@4.2.4

## 4.2.1

### Patch Changes

- Updated dependencies [cc2f87a8]
  - @linaria/utils@4.2.3
  - @linaria/tags@4.1.5

## 4.2.0

### Minor Changes

- 1e88e95d: Support for ECMAScript modules. Fixes #904 and #1043.

## 4.1.4

### Patch Changes

- 87ffe61c: The new `variableNameSlug` option that allows to customize css variable names (closes #1053).
- Updated dependencies [8a8be242]
- Updated dependencies [8a8be242]
- Updated dependencies [08304e09]
- Updated dependencies [87ffe61c]
  - @linaria/utils@4.2.2
  - @linaria/tags@4.1.4

## 4.1.3

### Patch Changes

- Updated dependencies [24b4a4bd]
  - @linaria/utils@4.2.1
  - @linaria/tags@4.1.3

## 4.1.2

### Patch Changes

- Updated dependencies [8590e134]
- Updated dependencies [f7351b09]
- Updated dependencies [c0bd271a]
- Updated dependencies [8f90fa75]
- Updated dependencies [ac0991a6]
  - @linaria/utils@4.2.0
  - @linaria/tags@4.1.2

## 4.1.1

### Patch Changes

- Updated dependencies [50bc0c79]
  - @linaria/utils@4.1.1
  - @linaria/tags@4.1.1

## 4.1.0

### Patch Changes

- Updated dependencies [92f6d871]
  - @linaria/utils@4.1.0
  - @linaria/tags@4.1.0

## 4.0.0

### Major Changes

- bc0cbeea: A completely new async mode with native support for Vite, Rollup, esbuild and Webpack resolvers.

  BREAKING CHANGES: Despite the fact, that it should be fully compatible with 3.0 and 2.0 branches, the new version of styles evaluator can have some serious bugs which can make your project unbuildable (however, since there is no runtime, if the build is finished successfully, everything will continue work as it was on 2.0 and 3.0). If you face some problems please let us know and we will fix it as soon as possible.

### Patch Changes

- 8be5650d: The repo has been migrated to PNPM and Turborepo
- ea41d440: New package @linaria/tags that contains all abstract logic for tags processors.
- 9a50c1c1: Linaria now removes all unused css-related code from the runtime.
- 4cdf0315: Tagged template-specific logic has been moved from `BaseProcessor` to `TaggedTemplateProcessor`. `BaseProcessor` now can be used to define any type of expressions for zero-runtime transformations, such as `makeStyles` from `@griffel/react`.
- 12d35cb9: `processors` aliases have been lost during publishing. (fixes #984)
- f726eb3b: `cx` should return `LinariaClassName` if all arguments are `LinariaClassName`
- 17c83e34: Aliases for environments without the support of `exports` in package.json.
- f0cddda4: Extends `BaseProcessor` to support tags other than tagged templates, such as `makeStyles` from `@griffel/react`.
- Updated dependencies [f0cddda4]
  - @linaria/logger@4.0.0
  - @linaria/utils@4.0.0
  - @linaria/tags@4.0.0

## 3.0.0-beta.21

### Patch Changes

- f726eb3b: `cx` should return `LinariaClassName` if all arguments are `LinariaClassName`
- 17c83e34: Aliases for environments without the support of `exports` in package.json.

## 3.0.0-beta.20

### Patch Changes

- 8be5650d: The repo has been migrated to PNPM and Turborepo
- Updated dependencies [8be5650d]
  - @linaria/logger@3.0.0-beta.20
  - @linaria/utils@3.0.0-beta.20

# [3.0.0-beta.19](https://github.com/callstack/linaria/compare/v3.0.0-beta.18...v3.0.0-beta.19) (2022-06-03)

### Features

- **babel:** api for custom tags ([#976](https://github.com/callstack/linaria/issues/976)) ([3285ccc](https://github.com/callstack/linaria/commit/3285ccc1d00449b78b3fc74087528cd38cbdd116))
- **babel:** new way for detecting tag imports ([#974](https://github.com/callstack/linaria/issues/974)) ([3305cfb](https://github.com/callstack/linaria/commit/3305cfb0c0f65abdacceeb7e6bad118c59f7d551))

# [3.0.0-beta.18](https://github.com/callstack/linaria/compare/v3.0.0-beta.17...v3.0.0-beta.18) (2022-04-01)

**Note:** Version bump only for package @linaria/core

# [3.0.0-beta.15](https://github.com/callstack/linaria/compare/v3.0.0-beta.14...v3.0.0-beta.15) (2021-11-29)

### Features

- **atomic:** create an atomic package for the css API ([#867](https://github.com/callstack/linaria/issues/867)) ([4773bcf](https://github.com/callstack/linaria/commit/4773bcf4b14f08cdc4d2b612654b962cdfc97eaa))

# [3.0.0-beta.13](https://github.com/callstack/linaria/compare/v3.0.0-beta.12...v3.0.0-beta.13) (2021-09-13)

### Bug Fixes

- **core:** return type alias instead of string from `css` and `cx` ([#835](https://github.com/callstack/linaria/issues/835)) ([7eb9d94](https://github.com/callstack/linaria/commit/7eb9d94dc2d9d79f7be0159c43fa5d71c96d7182))

# [3.0.0-beta.4](https://github.com/callstack/linaria/compare/v3.0.0-beta.3...v3.0.0-beta.4) (2021-05-07)

**Note:** Version bump only for package @linaria/core

# [3.0.0-beta.3](https://github.com/callstack/linaria/compare/v3.0.0-beta.2...v3.0.0-beta.3) (2021-04-20)

**Note:** Version bump only for package @linaria/core

# [3.0.0-beta.2](https://github.com/callstack/linaria/compare/v3.0.0-beta.1...v3.0.0-beta.2) (2021-04-11)

### Bug Fixes

- **core:** remove unnecessary spread operators from css and cx ([#746](https://github.com/callstack/linaria/issues/746)) ([#749](https://github.com/callstack/linaria/issues/749)) ([de23a09](https://github.com/callstack/linaria/commit/de23a0926c2583db01e7df5ea9a134f5910f96a1))
