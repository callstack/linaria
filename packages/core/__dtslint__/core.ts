/* eslint-disable @typescript-eslint/no-unused-vars */
import { css, cx } from '../src';

// $ExpectType LinariaClassName
const class1 = css``;

const activeClass = css``;

// $ExpectType string
const combined1 = cx(class1, 'active');

// $ExpectType LinariaClassName
const combined2 = cx(class1, activeClass);
