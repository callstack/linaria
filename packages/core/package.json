{
  "name": "@linaria/core",
  "version": "4.2.6",
  "description": "Blazing fast zero-runtime CSS in JS library",
  "keywords": [
    "css",
    "css-in-js",
    "linaria",
    "react",
    "styled-components"
  ],
  "homepage": "https://github.com/callstack/linaria#readme",
  "bugs": "https://github.com/callstack/linaria/issues",
  "repository": "git@github.com:callstack/linaria.git",
  "license": "MIT",
  "sideEffects": false,
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "types": "./types/index.d.ts",
      "import": "./dist/index.mjs",
      "default": "./dist/index.js"
    },
    "./*": {
      "types": "./types/*.d.ts",
      "import": "./dist/*.mjs",
      "default": "./dist/*.js"
    }
  },
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "types": "types/index.d.ts",
  "typesVersions": {
    "*": {
      "processors/*": [
        "./types/processors/*.d.ts"
      ]
    }
  },
  "files": [
    "dist/",
    "processors/",
    "types/"
  ],
  "linaria": {
    "tags": {
      "css": "./dist/processors/css.js"
    }
  },
  "scripts": {
    "build": "pnpm build:dist && pnpm build:declarations",
    "build:corejs-test": "cross-env NODE_ENV=legacy babel src --out-dir lib --extensions '.js,.jsx,.ts,.tsx' --ignore \"src/processors/**/*\"",
    "build:declarations": "tsc --emitDeclarationOnly --outDir types",
    "build:dist": "tsup --format cjs,esm",
    "test": "jest --config ../../jest.config.js --rootDir .",
    "test:dts": "dtslint --localTs ../../node_modules/typescript/lib __dtslint__",
    "typecheck": "tsc --noEmit --composite false",
    "watch": "pnpm build:dist --watch & pnpm build:declarations --watch"
  },
  "dependencies": {
    "@linaria/logger": "workspace:^",
    "@linaria/tags": "workspace:^",
    "@linaria/utils": "workspace:^"
  },
  "devDependencies": {
    "@babel/traverse": "^7.20.1",
    "@babel/types": "^7.20.2",
    "@types/babel__core": "^7.1.19",
    "@types/babel__traverse": "^7.17.1",
    "@types/node": "^17.0.39"
  },
  "engines": {
    "node": "^12.16.0 || >=13.7.0"
  },
  "publishConfig": {
    "access": "public"
  },
  "tsup": {
    "entry": [
      "src/index.ts",
      "src/processors/css.ts"
    ],
    "splitting": false,
    "sourcemap": true,
    "clean": true
  }
}
