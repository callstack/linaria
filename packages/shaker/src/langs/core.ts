import { types as t } from '@babel/core';
import type {
  AssignmentExpression,
  Block,
  CallExpression,
  Directive,
  ExpressionStatement,
  ForInStatement,
  ForStatement,
  Function,
  Identifier,
  IfStatement,
  MemberExpression,
  Node,
  ObjectExpression,
  SequenceExpression,
  SwitchCase,
  SwitchStatement,
  Terminatorless,
  TryStatement,
  VariableDeclaration,
  VariableDeclarator,
  WhileStatement,
} from '@babel/types';

import { peek } from '@linaria/babel-preset';
import type { IdentifierHandlers, Visitors } from '../types';
import GraphBuilderState from '../GraphBuilderState';
import ScopeManager from '../scope';
import DepsGraph from '../DepsGraph';

function isIdentifier(
  node: Node,
  name?: string | string[]
): node is Identifier {
  return (
    t.isIdentifier(node) &&
    (name === undefined ||
      (Array.isArray(name) ? name.includes(node.name) : node.name === name))
  );
}

type SideEffect = [
  {
    callee?: (child: CallExpression['callee']) => boolean;
    arguments?: (child: CallExpression['arguments']) => boolean;
  },
  (node: CallExpression, state: GraphBuilderState) => void
];

const sideEffects: SideEffect[] = [
  [
    // if the first argument of forEach is required, mark forEach as required
    {
      callee: (node) =>
        t.isMemberExpression(node) &&
        t.isIdentifier(node.property) &&
        node.property.name === 'forEach',
    },
    (node, state) => state.graph.addEdge(node.arguments[0], node),
  ],
];

function getCallee(node: CallExpression): Node {
  if (
    t.isSequenceExpression(node.callee) &&
    node.callee.expressions.length === 2
  ) {
    const [first, second] = node.callee.expressions;
    if (t.isNumericLiteral(first) && first.value === 0) {
      return second;
    }
  }

  return node.callee;
}

function findWildcardReexportStatement(
  node: t.CallExpression,
  identifierName: string,
  graph: DepsGraph
): t.Statement | null {
  if (!t.isIdentifier(node.callee) || node.callee.name !== 'require')
    return null;

  const declarator = graph.getParent(node);
  if (!t.isVariableDeclarator(declarator)) return null;

  const declaration = graph.getParent(declarator);
  if (!t.isVariableDeclaration(declaration)) return null;

  const program = graph.getParent(declaration);
  if (!t.isProgram(program)) return null;

  // Our node is a correct export
  // Let's check that we have something that looks like transpiled re-export
  return (
    program.body.find((statement) => {
      /*
       * We are looking for `Object.keys(_bar).forEach(…)`
       */

      if (!t.isExpressionStatement(statement)) return false;

      const expression = statement.expression;
      if (!t.isCallExpression(expression)) return false;

      const callee = expression.callee;
      if (!t.isMemberExpression(callee)) return false;

      const { object, property } = callee;

      if (!isIdentifier(property, 'forEach')) return false;

      if (!t.isCallExpression(object)) return false;

      // `object` should be `Object.keys`
      if (
        !t.isMemberExpression(object.callee) ||
        !isIdentifier(object.callee.object, 'Object') ||
        !isIdentifier(object.callee.property, 'keys')
      )
        return false;

      //
      const [argument] = object.arguments;
      return isIdentifier(argument, identifierName);
    }) ?? null
  );
}

/*
 * Returns nodes which are implicitly affected by specified node
 */
function getAffectedNodes(node: Node, state: GraphBuilderState): Node[] {
  // FIXME: this method should be generalized
  const callee = t.isCallExpression(node) ? getCallee(node) : null;
  if (
    t.isCallExpression(node) &&
    t.isMemberExpression(callee) &&
    isIdentifier(callee.object, 'Object') &&
    isIdentifier(callee.property, [
      'assign',
      'defineProperty',
      'defineProperties',
      'freeze',
      'observe',
    ])
  ) {
    const [obj, property] = node.arguments;
    if (!t.isIdentifier(obj)) {
      return [];
    }

    if (
      state.scope.getDeclaration(obj) !== ScopeManager.globalExportsIdentifier
    ) {
      return [node.arguments[0]];
    }

    if (t.isStringLiteral(property)) {
      if (property.value === '__esModule') {
        return [node.arguments[0]];
      }

      state.graph.addExport(property.value, node);
    }
  }

  return [];
}

export const visitors: Visitors = {
  /*
   * FunctionDeclaration | FunctionExpression | ObjectMethod | ArrowFunctionExpression | ClassMethod | ClassPrivateMethod;
   * Functions can be either a statement or an expression.
   * That's why we need to disable default dependency resolving strategy for expressions by passing `ignoreDeps` flag.
   * Every function must have a body. Without a body, it becomes invalid.
   * In general, a body depends on parameters of a function.
   * In real life, some of the parameters can be omitted, but it's not trivial to implement that type of tree shaking.
   */
  Function(this: GraphBuilderState, node: Function) {
    const unsubscribe = this.onVisit((descendant) =>
      this.graph.addEdge(node, descendant)
    );
    this.baseVisit(node, true); // ignoreDeps=true prevents default dependency resolving
    unsubscribe();

    this.graph.addEdge(node, node.body);

    node.params.forEach((param) => this.graph.addEdge(node.body, param));
    if (
      t.isFunctionExpression(node) &&
      node.id !== null &&
      node.id !== undefined
    ) {
      // keep function name in expressions like `const a = function a();`
      this.graph.addEdge(node, node.id);
    }
  },

  /*
   * ExpressionStatement
   */
  ExpressionStatement(this: GraphBuilderState, node: ExpressionStatement) {
    this.baseVisit(node);

    this.graph.addEdge(node, node.expression);
    this.graph.addEdge(node.expression, node);
  },

  /*
   * BlockStatement | Program
   * The same situation as in ExpressionStatement: if one of the expressions is required, the block itself is also required.
   * Whereas a block doesn't depend on its children.
   * Example:
   * 1. let c;
   * 2. { // BlockStatement begin
   * 3.   let a = 1;
   * 4.   let b = 2;
   * 5.   a++;
   * 6.   a = c;
   * 7. } // BlockStatement end
   *
   * If we want to evaluate the value of `c`, we need to evaluate lines 1, 3, 5 and 6,
   * but we don't need line 4, even though it's a child of the block.
   */
  Block(this: GraphBuilderState, node: Block) {
    this.baseVisit(node);

    if (t.isProgram(node)) {
      const exportsDeclaration = this.scope.getDeclaration('global:exports')!;
      this.graph.addEdge(node, exportsDeclaration);
      node.directives.forEach((directive) =>
        this.graph.addEdge(node, directive)
      );
    }
  },

  Directive(this: GraphBuilderState, node: Directive) {
    this.baseVisit(node);
    this.graph.addEdge(node, node.value);
  },

  /*
   * TryStatement
   * try { /* block *\/ } catch() {/* handler *\/} finalize {/* finalizer *\/}
   * `handler` and `finalizer` do not make sense without `block`
   * `block` depends on the whole node.
   */
  TryStatement(this: GraphBuilderState, node: TryStatement) {
    this.baseVisit(node);
    [node.handler, node.finalizer].forEach((statement) => {
      if (statement) {
        this.graph.addEdge(node.block, statement);
        this.graph.addEdge(statement, node.block);
      }
    });
  },

  IfStatement(this: GraphBuilderState, node: IfStatement) {
    this.baseVisit(node);
    this.graph.addEdge(node, node.consequent);
    this.graph.addEdge(node, node.test);
  },

  /*
   * WhileStatement
   * Pretty simple behaviour here:
   * • if body is required, the statement is required
   * • if the statement is required, the condition is also required.
   */
  WhileStatement(this: GraphBuilderState, node: WhileStatement) {
    this.baseVisit(node);
    this.graph.addEdge(node, node.test);
  },

  SwitchCase(this: GraphBuilderState, node: SwitchCase) {
    this.baseVisit(node);
    node.consequent.forEach((statement) => this.graph.addEdge(statement, node));
    if (node.test) {
      this.graph.addEdge(node, node.test);
    }
  },

  SwitchStatement(this: GraphBuilderState, node: SwitchStatement) {
    this.baseVisit(node);
    node.cases.forEach((c) => this.graph.addEdge(c, node));
    this.graph.addEdge(node, node.discriminant);
  },

  ForStatement(this: GraphBuilderState, node: ForStatement) {
    this.baseVisit(node);

    [node.init, node.test, node.update, node.body].forEach((child) => {
      if (child) {
        this.graph.addEdge(node, child);
      }
    });
  },

  /*
   * ForInStatement
   * for (const k in o) { body }
   */
  ForInStatement(this: GraphBuilderState, node: ForInStatement) {
    this.baseVisit(node);

    if (node.body) {
      this.graph.addEdge(node, node.body);
      this.graph.addEdge(node.body, node.left);
    }

    this.graph.addEdge(node.left, node.right);
  },

  /*
   * BreakStatement | ContinueStatement | ReturnStatement | ThrowStatement | YieldExpression | AwaitExpression
   * All these nodes are required to evaluate the value of a function in which they are defined.
   * Also, the value of these nodes depends on the argument if it is presented.
   */
  Terminatorless(this: GraphBuilderState, node: Terminatorless) {
    this.baseVisit(node);

    if (
      !(t.isBreakStatement(node) || t.isContinueStatement(node)) &&
      node.argument
    ) {
      this.graph.addEdge(node, node.argument);
    }

    const closestFunctionNode = peek(this.fnStack);
    this.graph.addEdge(closestFunctionNode, node);
  },

  /*
   * ObjectExpression
   * Objects are… complicated. Especially because similarly looking code can be either an expression or a pattern.
   * In this case we work with an expression like:
   * const obj = {
   *   method() {}, // ObjectMethod
   *   property: "value", // ObjectProperty
   *   ...rest, // SpreadElement
   * }
   */
  ObjectExpression(this: GraphBuilderState, node: ObjectExpression) {
    this.context.push('expression');
    this.baseVisit(node);
    node.properties.forEach((prop) => {
      this.graph.addEdge(node, prop);
      if (t.isObjectMethod(prop)) {
        this.graph.addEdge(prop, prop.key);
        this.graph.addEdge(prop, prop.body);
      } else if (t.isObjectProperty(prop)) {
        this.graph.addEdge(prop, prop.key);
        this.graph.addEdge(prop, prop.value);
      } else if (t.isSpreadElement(prop)) {
        this.graph.addEdge(prop, prop.argument);
      }
    });
    this.context.pop();
  },

  /*
   * MemberExpression
   * It's about a simple expression like `obj.foo` or `obj['foo']`.
   * In addition to default behaviour (an expression depends on all its children),
   * we add a backward dependency from an object to a node for processing member
   * expressions in assignments.
   *
   * Example:
   * let obj = { a: 1 };
   * obj.b = 2;
   *
   * If we try to evaluate `obj` without backward dependency,
   * `obj.b = 2` will be cut and we will get just `{ a: 1 }`.
   */
  MemberExpression(this: GraphBuilderState, node: MemberExpression) {
    this.baseVisit(node);

    if (
      isIdentifier(node.object, 'exports') &&
      this.scope.getDeclaration(node.object) ===
        ScopeManager.globalExportsIdentifier
    ) {
      // We treat `exports.something` and `exports['something']` as identifiers in the global scope
      this.graph.addEdge(node, node.object);
      this.graph.addEdge(node, node.property);

      const isLVal = peek(this.context) === 'lval';
      if (isLVal) {
        this.scope.declare(node, false);
      } else {
        const declaration = this.scope.addReference(node);
        this.graph.addEdge(node, declaration);
      }

      return;
    }

    if (t.isIdentifier(node.object) && t.isIdentifier(node.property)) {
      // It's simple `foo.bar` expression. Is it a usage of a required library?
      const declaration = this.scope.getDeclaration(node.object);
      if (
        t.isIdentifier(declaration) &&
        this.graph.importAliases.has(declaration) &&
        !node.computed
      ) {
        // It is. We can remember what exactly we use from it.
        const source = this.graph.importAliases.get(declaration)!;
        this.graph.imports.get(source)!.push(node.property);
      }
    }
  },

  /*
   * AssignmentExpression
   * `a = b`, `{ ...rest } = obj`, `obj.a = 3`, etc.
   * It's not a declaration, it's just an assignment, but it affects
   * the value of declared variable if the variable it mentioned in the left part.
   * So, we apply some context-magic here in order to catch reference of variables in the left part.
   * We switch the context to `lval` and continue traversing through the left branch.
   * If we then meet some identifier, we mark it as a dependency of its declaration.
   */
  AssignmentExpression(this: GraphBuilderState, node: AssignmentExpression) {
    this.context.push('lval');
    this.visit<AssignmentExpression['left'], AssignmentExpression>(
      node.left,
      node,
      'left'
    );
    this.context.pop();

    this.visit(node.right, node, 'right');

    // The value of an expression depends on the left part.
    this.graph.addEdge(node, node.left);

    // The left part of an assignment depends on the right part.
    this.graph.addEdge(node.left, node.right);
  },

  /*
   * VariableDeclarator
   * It would be pretty simple if it weren't used to declare variables from other modules.
   */
  VariableDeclarator(this: GraphBuilderState, node: VariableDeclarator) {
    /*
     * declared is used for detecting external dependencies in cases like
     * const { a, b, c } = require('module');
     *
     * We are remembering all declared variables in order to use it later in CallExpression visitor
     */
    const declared: Array<[Identifier, Identifier | null]> = [];
    this.meta.set('declared', declared);
    const unregister = this.scope.addDeclareHandler((identifier, from) =>
      declared.push([identifier, from])
    );
    this.baseVisit(node);
    this.meta.delete('declared');
    unregister();

    if (node.init) {
      // If there is an initialization part, the identifier depends on it.
      this.graph.addEdge(node.id, node.init);
    }

    // If a statement is required itself, an id is also required
    this.graph.addEdge(node, node.id);
  },

  /*
   * VariableDeclaration
   * It's just a wrapper for group of VariableDeclarator.
   * If one of the declarators is required, the wrapper itself is also required.
   */
  VariableDeclaration(this: GraphBuilderState, node: VariableDeclaration) {
    this.meta.set('kind-of-declaration', node.kind);
    this.baseVisit(node);
    node.declarations.forEach((declaration) =>
      this.graph.addEdge(declaration, node)
    );
    this.meta.delete('kind-of-declaration');
  },

  /*
   * CallExpression
   * Do you remember that we have already mentioned it in VariableDeclarator?
   * It is a simple expression with default behaviour unless it is a `require`.
   *
   * Another tricky use case here is functions with side effects (e.g. `Object.defineProperty`).
   */
  CallExpression(
    this: GraphBuilderState,
    node: CallExpression,
    parent: Node | null
  ) {
    this.baseVisit(node);

    if (t.isIdentifier(node.callee) && node.callee.name === 'require') {
      // It looks like a module import …
      const scopeId = this.scope.whereIsDeclared(node.callee);
      if (scopeId && scopeId !== 'global') {
        // … but it is just a user defined function
        return;
      }

      const [firstArg] = node.arguments;
      if (!t.isStringLiteral(firstArg)) {
        // dynamic import? Maybe someday we can do something about it
        return;
      }

      const { value: source } = firstArg;
      const declared = this.meta.get('declared') as Array<
        [Identifier, Identifier | null]
      >;
      if (!declared) {
        // This is a standalone `require`
        return;
      }

      // Define all declared variables as external dependencies.
      declared.forEach(([local, _imported]) =>
        // FIXME: var slugify = require('../slugify').default;
        {
          if (!this.graph.imports.has(source)) {
            this.graph.imports.set(source, []);
          }

          if (
            parent &&
            t.isMemberExpression(parent) &&
            t.isIdentifier(parent.property)
          ) {
            // An imported function is specified right here.
            // eg. require('../slugify').default
            this.graph.imports.get(source)!.push(parent.property);
          } else {
            if (
              t.isCallExpression(parent) &&
              t.isIdentifier(parent.callee) &&
              typeof parent.callee.name === 'string'
            ) {
              if (parent.callee.name.startsWith('_interopRequireDefault')) {
                this.graph.importTypes.set(source, 'default');
              } else if (
                parent.callee.name.startsWith('_interopRequireWildcard')
              ) {
                this.graph.importTypes.set(source, 'wildcard');
              } else {
                // What I've missed?
              }
            }

            // Do we know the type of import?
            if (!this.graph.importTypes.has(source)) {
              // Is it a wildcard reexport? Let's check.
              const statement = findWildcardReexportStatement(
                node,
                local.name,
                this.graph
              );
              if (statement) {
                this.graph.addEdge(local, statement);
                this.graph.reexports.push(local);
                this.graph.importTypes.set(source, 'reexport');
              }
            }

            // The whole namespace was imported. We will know later, what exactly we need.
            // eg. const slugify = require('../slugify');
            this.graph.importAliases.set(local, source);
          }
        }
      );

      return;
    }

    sideEffects.forEach(([conditions, callback]) => {
      if (
        (conditions.callee && !conditions.callee(node.callee)) ||
        (conditions.arguments && !conditions.arguments(node.arguments))
      ) {
        return;
      }

      return callback(node, this);
    });

    getAffectedNodes(node, this).forEach((affectedNode) => {
      this.graph.addEdge(affectedNode, node);
      if (t.isIdentifier(affectedNode)) {
        this.graph.addEdge(
          this.scope.getDeclaration(affectedNode)!,
          affectedNode
        );
      }
    });
  },

  /*
   * SequenceExpression
   * It is a special case of expression in which the value of the whole
   * expression depends only on the last subexpression in the list.
   * The rest of the subexpressions can be omitted if they don't have dependent nodes.
   *
   * Example:
   * const a = (1, 2, b = 3, 4, b + 2); // `a` will be equal 5
   */
  SequenceExpression(this: GraphBuilderState, node: SequenceExpression) {
    // Sequence value depends on only last expression in the list
    this.baseVisit(node, true);
    if (node.expressions.length > 0) {
      this.graph.addEdge(node, node.expressions[node.expressions.length - 1]);
    }
  },
};

export const identifierHandlers: IdentifierHandlers = {
  declare: [
    ['CatchClause', 'param'],
    ['Function', 'params'],
    ['FunctionExpression', 'id'],
    ['RestElement', 'argument'],
    ['ThrowStatement', 'argument'],
    ['VariableDeclarator', 'id'],
  ],
  keep: [['ObjectProperty', 'key']],
  refer: [
    ['ArrayExpression', 'elements'],
    ['AssignmentExpression', 'left', 'right'],
    ['BinaryExpression', 'left', 'right'],
    ['CallExpression', 'arguments', 'callee'],
    ['ConditionalExpression', 'test', 'consequent', 'alternate'],
    ['ForInStatement', 'right'],
    ['Function', 'body'],
    ['IfStatement', 'test'],
    ['LogicalExpression', 'left', 'right'],
    ['NewExpression', 'arguments', 'callee'],
    ['ObjectProperty', 'value'],
    ['ReturnStatement', 'argument'],
    ['SequenceExpression', 'expressions'],
    ['SwitchStatement', 'discriminant'],
    ['UnaryExpression', 'argument'],
    ['UpdateExpression', 'argument'],
    ['VariableDeclarator', 'init'],
  ],
};
