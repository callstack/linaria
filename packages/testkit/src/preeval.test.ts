import { join } from 'path';

import { transformSync } from '@babel/core';
import dedent from 'dedent';

import { preeval } from '@linaria/babel-preset';

const run = (code: TemplateStringsArray) => {
  const filename = join(__dirname, 'source.ts');
  const formattedCode = dedent(code);

  const transformed = transformSync(formattedCode, {
    babelrc: false,
    configFile: false,
    filename,
    plugins: [
      '@babel/plugin-syntax-typescript',
      [
        preeval,
        {
          evaluate: true,
        },
      ],
    ],
  });

  if (!transformed) {
    throw new Error(`Something went wrong with ${filename}`);
  }

  return {
    code: transformed.code,
    // metadata: transformed.metadata.__linariaShaker,
  };
};

describe('preeval', () => {
  it('should keep getGlobal but remove window-related code', () => {
    const { code } = run`
      function getGlobal() {
        if (typeof globalThis !== "undefined") {
          return globalThis;
        }

        if (typeof window !== "undefined") {
          return window;
        }

        if (typeof global !== "undefined") {
          return global;
        }

        if (typeof self !== "undefined") {
          return self;
        }

        return mockGlobal;
      }
    `;

    expect(code).toMatchSnapshot();
  });

  it('should remove usages of window scoped identifiers', () => {
    const { code } = run`
      $RefreshReg$("Container");
      if (import.meta.hot) {
        window.$RefreshReg$ = () => {};
      }

      $RefreshReg$("Header");
    `;

    expect(code).toMatchSnapshot();
  });

  it('should not remove "location" in types only because it looks like a global variable', () => {
    const { code } = run`
      interface IProps {
        fn: (location: string) => void;
      }
    `;

    expect(code).toMatchSnapshot();
  });

  it('should keep object members that look like window globals', () => {
    const { code } = run`
      class Test {
        fetch: typeof global.fetch;
        constructor(options) {
          this.fetch = options.fetch;
        }
      }
    `;

    expect(code).toMatchSnapshot();
  });

  it('should keep type parameters that look like window globals', () => {
    const { code } = run`
      const blah = window.Foo;
      type FooType = Generic<Foo>;
    `;

    expect(code).toMatchSnapshot();
  });
});
