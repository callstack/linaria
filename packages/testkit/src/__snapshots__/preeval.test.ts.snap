// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preeval should keep getGlobal but remove window-related code 1`] = `
"function getGlobal() {
  if (typeof globalThis !== \\"undefined\\") {
    return globalThis;
  }
  if (typeof global !== \\"undefined\\") {
    return global;
  }
  if (typeof self !== \\"undefined\\") {
    return self;
  }
  return mockGlobal;
}"
`;

exports[`preeval should keep object members that look like window globals 1`] = `
"class Test {
  fetch: typeof global.fetch;
  constructor(options) {
    this.fetch = options.fetch;
  }
}"
`;

exports[`preeval should keep type parameters that look like window globals 1`] = `"type FooType = Generic<Foo>;"`;

exports[`preeval should not remove "location" in types only because it looks like a global variable 1`] = `
"interface IProps {
  fn: (location: string) => void;
}"
`;

exports[`preeval should remove usages of window scoped identifiers 1`] = `""`;
